### PR TITLE
Disable http client connection reuse to prevent memory leak

### LINKED ISSUE #700: [BUG] Memory Leak in Operator version 2.4.0
### What is the bug?
After upgrading to Opensearch Operator 2.4.0 (from 2.3.2) we are experiencing ever-increasing memory usage and OOM killings of the operator pod. This appears to be escalated further when applying Opensearch resources (OpensearchUser, OpensearchRole and OpensearchRoleBindings)

### How can one reproduce the bug?
Deploy the operator 2.4.0 and install a opensearch cluster. We are currently using OpenSearch version 2.7.0

### What is the expected behavior?
Memory consumption of the Operator pod should be fairly stable. 

### What is your host/environment?
* Azure Kubernetes cluster running Kubernetes version 1.26.6
* OpenSearch k8s Operator 2.4.0

### Do you have any screenshots?
Screenshot of memory usage by the operator before and after upgrade. Upgrade was performed January 9th and at that time the memory consumption started to increase until the pod was OOM-killed. 

Screenshot 2024-01-17 at 12 16 37

### Do you have any additional context?
We have seen the same memory consumption pattern on 5 different kubernetes clusters, before update of operator the memory consumption was pretty stable. When the upgrade was performed memory consumption started to increase. The rate of which consumption increases seems to be related to the amount of Custom Opensearch Resources that has been deployed. 

It appears that the same was seen in https://github.com/opensearch-project/opensearch-k8s-operator/issues/612 but here no real solution was found/proposed, however we see memory leak constant on all clusters. 

Memory increases more rapidly if we deploy more resources like this (replacing demo2 with demoXX). 
```---
apiVersion: opensearch.opster.io/v1
kind: OpensearchUser
metadata:
  name: demo2
  namespace: opensearch
spec:
  opensearchCluster:
    name: test-opensearch
  passwordFrom:
    name: opensearch-default-userpass
    key: password
  backendRoles:
    - opensearch-demo2-read
    - opensearch-demo2-write
---
apiVersion: opensearch.opster.io/v1
kind: OpensearchRole
metadata:
  name: opensearch-demo2-read
  namespace: opensearch
spec:
  opensearchCluster:
    name: test-opensearch
  clusterPermissions:
    - "cluster:monitor/nodes/info"
    - "cluster:monitor/health"
    - "indices:admin/template/get"
    - "indices:admin/exists"
  indexPermissions:
    - indexPatterns:
        - demo2*
      allowedActions:
        - read_access
---
apiVersion: opensearch.opster.io/v1
kind: OpensearchRole
metadata:
  name: opensearch-demo2-write
  namespace: opensearch
spec:
  opensearchCluster:
    name: test-opensearch
  clusterPermissions:
    - "cluster:monitor/nodes/info"
    - "cluster:monitor/health"
    - "indices:admin/template/get"
    - "indices:admin/exists"
    - "indices:admin/delete"
  indexPermissions:
    - indexPatterns:
        - demo2*
      allowedActions:
        - write_access
---        
apiVersion: opensearch.opster.io/v1
kind: OpensearchUserRoleBinding
metadata:
  name: demo2
  namespace: opensearch
spec:
  opensearchCluster:
    name: test-opensearch
  roles:
  - opensearch-demo2-read
  - opensearch-demo2-write
  users:
  - demo2
```

### PR DESCRIPTION
**Pull request by swoehrl-mw**
### Description
The operator pod is suffering from memory leaks. After some analysis I think I have narrowed it down to connections for the http client being kept for reuse but never being used due to a new client being created in every reconcile run.
This PR disables the connection keepalive/reuse and (at least in my experiments) prevents the memory leak.

### Issues Resolved
Fixes #700 

### Check List
- [x] Commits are signed per the DCO using --signoff 
- [-] Unittest added for the new/changed functionality and all unit tests are successful
- [-] Customer-visible features documented
- [x] No linter warnings (`make lint`)

If CRDs are changed:
- [-] CRD YAMLs updated (`make manifests`) and also copied into the helm chart
- [-] Changes to CRDs documented

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
Thanks @swoehrl-mw I will take a look at this today.

File: opensearch-operator/pkg/reconcilers/scaler.go [hunk 1]
```diff
@@ -33,14 +35,17 @@ func NewScalerReconciler(
 	recorder record.EventRecorder,
 	reconcilerContext *ReconcilerContext,
 	instance *opsterv1.OpenSearchCluster,
-	opts ...reconciler.ResourceReconcilerOption,
+	opts ...ReconcilerOption,
 ) *ScalerReconciler {
```
Comment (prudhvigodithi) [inline opensearch-operator/pkg/reconcilers/scaler.go:39]: May I know what is change doing by adding `ReconcilerOption` ?

File: opensearch-operator/pkg/reconcilers/scaler.go [hunk 2]
```diff
@@ -187,11 +192,7 @@ func (r *ScalerReconciler) decreaseOneNode(currentStatus opsterv1.ComponentStatu
 	if !smartDecrease {
 		return false, err
 	}
-	username, password, err := helpers.UsernameAndPassword(r.client, r.instance)
-	if err != nil {
-		return true, err
-	}
-	clusterClient, err := services.NewOsClusterClient(builders.URLForCluster(r.instance), username, password)
+	clusterClient, err := util.CreateClientForCluster(r.client, r.ctx, r.instance, r.osClientTransport)
```
Comment (prudhvigodithi) [inline opensearch-operator/pkg/reconcilers/scaler.go:195]: Also what is the advantage of switching this to `CreateClientForCluster` method?

Comment (prudhvigodithi) [reply opensearch-operator/pkg/reconcilers/scaler.go:195]: I see the following code also changes to `CreateClientForCluster`, what is this change doing ?

**Comment by prudhvigodithi** [comment]
Thanks @swoehrl-mw I have added my comments can you please check, this is an important change that should be shipped.
Adding @salyh 
@getsaurabh02

File: opensearch-operator/pkg/reconcilers/scaler.go (hunk 1 above)
Comment (swoehrl-mw) [reply opensearch-operator/pkg/reconcilers/scaler.go:39]: That change is needed to, for one, make this reconciler behave more like the others, and also to get the osClientTransport field from the ReconcilerOptions which is a needed argument for the `CreateClientForCluster` function. In the end this also helps with testing as the RoundTripper can be mocked.

File: opensearch-operator/pkg/reconcilers/scaler.go (hunk 2 above)
Comment (swoehrl-mw) [reply opensearch-operator/pkg/reconcilers/scaler.go:195]: This basically unifies how an opensearch client can be created so there is no duplicated code and only one way to create a client.

Comment (prudhvigodithi) [reply opensearch-operator/pkg/reconcilers/scaler.go:195]: So with this only one OpenSearch client is created (until the operator is restarted) and used for all the operator options invoked like create/update on a cluster?

Comment (swoehrl-mw) [reply opensearch-operator/pkg/reconcilers/scaler.go:195]: A new client is created for each reconciler and reconcile loop, but there is only one codepath through which a client can be created. This gives us better control when we want/need to change handling of clients.
Strictly speaking this change is not needed for the bugfix, but as I was checking that code part anyway for the leak I used the chance to unify it.

**Review by prudhvigodithi — APPROVED** [review]

**Pull request merged**